---
title: "Hands-on session 4: cell type-level analysis in R"
author: "Ángeles Arzalluz-Luque"
date: "`r Sys.Date()`"
output:
  BiocStyle::html_document:
    toc: true
vignette: >
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Introduction

In this session, we will perform a series of analyses that can be performed
on already-clustered single-cell data, including:

- Assignment of cell labels using a reference dataset.
- Identification of marker genes for clusters for *de novo* cell label assignment.
- Differential expression analysis.
- Differential abundance analysis.

This tutorial is based on the following resources:

- [Azimuth tutorial](https://satijalab.github.io/azimuth/articles/run_azimuth_tutorial.html)
- "Orchestrating single-cell analysis with Bioconductor" book chapters:
  - [Marker gene detection](https://bioconductor.org/books/3.17/OSCA.basic/marker-detection.html)
  - [Cell type annotation](https://bioconductor.org/books/3.17/OSCA.basic/cell-type-annotation.html)
- [Seurat Differential Expression testing vignette](https://satijalab.org/seurat/articles/de_vignette)
- [Milo demo vignette](https://bioconductor.org/packages/release/bioc/vignettes/miloR/inst/doc/milo_demo.html)

## Required packages

We will first load a series of single-cell analysis R packages:

```{r, message=FALSE}
library(Seurat)
library(SeuratData)
library(SeuratDisk)
library(SingleCellExperiment)
library(Azimuth)
library(scran)
library(scater)
library(miloR)
```


## Data

All data and data-related files for this hands-on session will be available 
under the following directory:

```{r}
data_dir <- c("HandsOn4_CellType_Analysis/")
```

We will append this prefix to all the paths in the tutorial.

For this hands-on session, we will use the `SeuratObject` obtained after 
running all the analysis in hands-on session 3, stored as an H5Seurat file:

```{r}
setty_seurat <- LoadH5Seurat(paste0("../HandsOn3_SingleCell_Data_QC/", 
                                    "processed_data/setty_seurat_clustered.h5seurat"))
```


## Assignment of cell labels using a reference dataset.

After clustering cells, it might be difficult (or even risky) to assign cell
identities blindly. When well-curated references are available, cells can be
confidently and rapidly annotated by mapping to already-characterized cell 
atlases. 

The `Azimuth` R package provides not only functions, but also integrated
references for multiple human tissues, including bone marrow. These are
installed via the `SeuratData` package, which contains the following references: 

```{r}
available_data <- AvailableData()
available_data[grep("Azimuth", available_data[, 3]), 1:3]
```

We will use the `bonemarrowref` reference, but first we need to install and 
load the `SeuratData` package containing the bone marrow reference dataset 
using the source package that was previously downloaded:

```{r}
if(!"bonemarrowref.SeuratData" %in% installed.packages()[,"Package"]){
  install.packages("bonemarrowref.SeuratData_1.0.0.tar.gz",
                 repos = NULL, type = "source")
}

library(bonemarrowref.SeuratData)
```

To annotate cell types, we will use the `RunAzimuth()` function, which can 
take a `SeuratObject` as input and returns the same type of object including 
cell type annotations:

```{r, warning = FALSE}
setty_seurat <- RunAzimuth(query = setty_seurat,
                           reference = "bonemarrowref")
```

Let's explore the `RunAzimuth()` results:

```{r}
table(setty_seurat$predicted.celltype.l1)
table(setty_seurat$predicted.celltype.l2)
```

As you can see, the method provides cell type labels at two levels of hierarchy:
- There are four broad cell types, although most cells are classified as 
Hematopoietic Stem and Progenitor Cells (HSPCs).
- There are several sub-cell types within them, including different stages of cell
differentiation and lineages.

Let's visualize these results using tSNE:

```{r}
DimPlot(setty_seurat, group.by = "predicted.celltype.l1",
        reduction = "tsne", label = TRUE)
DimPlot(setty_seurat, group.by = "predicted.celltype.l2",
        reduction = "tsne", label = TRUE)
```


## Assignment of cell labels *de novo* using marker genes

Strategies for cell labelling are often based on the definition of **marker genes**, 
i.e. genes that are highly expressed in each cell type. This is useful when only the
identities of the markers are available.

First, to find the genes that are significantly highly expressed in each cell type,
we will use the `scoreMarkers()` function from the `scran` R package. This function
compares each pair of clusters and computes scores quantifying the differences 
in the expression distributions between clusters.

Importantly, `scran` belongs to the family of R packages that are based on 
the `SingleCellExperiment` data class, so we will need to transform our 
`SeuratObject` and correctly fill the slots with the normalized and 
log-transformed counts:

```{r}
setty_sce <- as.SingleCellExperiment(setty_seurat)
colLabels(setty_sce) <- setty_sce$ident
normcounts(setty_sce) <- setty_seurat[["RNA"]]@data
logcounts(setty_sce) <- log2(normcounts(setty_sce)+1)

table(setty_sce$predicted.celltype.l2)
```

Now, we will apply the `scoreMarkers()` function. We will define a minimum 
log fold-change of 1 to consider a gene as relevant to prevent small changes to
drive the detected markers:

```{r}
markers <- scoreMarkers(setty_sce, 
                        groups = colLabels(setty_sce),
                        lfc = 1)

colnames(markers[[1]])
```

The `markers` list contains one table per cluster including the marker scores
for each gene in that cluster. Since this is a very long list, and not all genes
will be relevant, we need to select one of these variables to **rank markers**
and subset the list.

A good metric to do this is **Area Under the Curve (AUC)**, which measures 
our ability to distinguish between two distributions in a pairwise comparison. 
The AUC represents the probability that a randomly chosen observation from our 
cluster of interest is greater than a randomly chosen observation from the other 
cluster. A value of 1 corresponds to upregulation, where all values of our cluster 
of interest are greater than any value from the other cluster; a value of 0.5 means 
that there is no net difference in the location of the distributions; and a 
value of 0 corresponds to downregulation.

Let's focus on **cluster 2**, and select the top 100 genes with the largest 
mean AUC:

```{r}
selected_cluster <- "2"

markers_select <- markers[[selected_cluster]]

top_markers <- order(markers_select$mean.AUC, 
                     decreasing = TRUE)[1:100]

markers_select[top_markers,1:4]
```

Let's plot the top markers in the cluster using the `plotExpression()` function in
the `scater` QC R package:

```{r}
plotExpression(setty_sce, 
               features = c("XACT", "ZNF385D", "CRACD", 
                            "ST8SIA6", "PDZD8"),
               x = "label", colour_by = "label")
```

While this is very interesting, gene names may not mean anything by themselves. 
To better interpret the **list of markers for cluster 2**, we will test these genes 
for enrichment of gene sets in the Gene Ontology (GO). Doing this, we will 
get a sense of the biological processes that are active in each cell type, 
which can give us clues regarding cell type identities.

There are many ways to do this, but in this case, we will use the `limma` R 
package and the `goana()` function, which performs a hypergeometric test to identify
GO terms that are overrepresented in our marker subset.

Since some of our gene IDs are in ENSEMBL format and others are gene symbols, 
and this function requires them to be Entrez IDs, we will map the identifiers 
to obtain the Entrez equivalents using the genome-wide human annotation:

```{r}
library(org.Hs.eg.db)

# map gene IDs to database
entrez_ids_ens <- mapIds(org.Hs.eg.db, 
                     keys = rownames(markers_select), 
                     column = "ENTREZID", keytype = "ENSEMBL")

entrez_ids_symb <- mapIds(org.Hs.eg.db, 
                     keys = rownames(markers_select), 
                     column = "ENTREZID", keytype = "SYMBOL")

# merge
entrez_ids_symb[is.na(entrez_ids_symb)] <- entrez_ids_ens[is.na(entrez_ids_symb)]
```

Now, we can run the `goana()` enrichment analysis on our top markers vs 
all markers:

```{r}
library(limma)

go_out <- goana(entrez_ids_symb[top_markers], 
                species = "Hs", 
                universe = entrez_ids_symb)

# Only keeping biological process terms that are not overly general
go_out <- go_out[order(go_out$P.DE),]
go_useful <- go_out[go_out$Ont=="BP" & go_out$N <= 150,]
head(go_useful[,c(1,3,4)], 40)
```

If we check the [GO database](https://geneontology.org/) for the most enriched 
term, "definitive hemopoiesis", we find that this corresponds to an intermediate 
state in erythrocyte differentiation. While other terms point to earlier stages 
in blood cell differentiation, such as "embryonic hemopoiesis" and "negative 
regulation of myeloid cell differentiation", terms such as "heme metabolic 
process" and "tetrapyrrole catabolic process" and "intracellular iron ion 
homeostasis" point to hemoglobin metabolism, suggesting that these cells are 
**erythrocytes or late-stage erythrocyte precursors**. If you go back to the
Azimuth cell type annotation results, you see that this cluster is in fact 
labelled as **early erythrocytes**.


## Differential Expression analysis

Even though marker gene detection is based on approaches that are similar to those
for Differential Expression (DE) analysis, we may want to obtain more detailed
results regarding the DE genes between two specific clusters. To achieve this,
we will use the `SeuratObject` and the Azimuth cell type identity results.

The bulk of Seurat’s differential expression features can be accessed through 
the `FindMarkers()` function. To test for differential expression between two 
specific groups of cells, we will specify the `ident.1` and `ident.2` parameters.
Let's perform a DE analysis comparing Common Lymphoid Progenitor (CLP) cells 
to a more differentiated stage, pre-B cells:


```{r}
table(setty_seurat$predicted.celltype.l2)

preB_markers <- FindMarkers(setty_seurat, 
                                     group.by = "predicted.celltype.l2",
                                     ident.1 = "pre B", 
                                     ident.2 = "CLP",
                                     logfc.threshold = 1)

head(preB_markers, 10)
```

See how T cell development markers CD44 is significantly more
highly expressed in CLP cells in comparison to pre-B cells.

The results data frame has the following columns :

- p_val : unadjusted p-value.
- avg_log2FC : log fold-change of the average expression between the two groups. 
Positive values indicate that the feature is more highly expressed in the first group.
- pct.1 : The percentage of cells where the feature is detected in the first group.
- pct.2 : The percentage of cells where the feature is detected in the second group.
- p_val_adj : Adjusted p-value, based on Bonferroni correction using all features in the dataset.

**NOTE**: this function can also be used to define **marker genes**, 
as we did in the previous section, however, it is based on a different approach:
instead of performing all pairwise comparisons, it compares cells from the selected
cell type to all the remaining cells. To do this, simply set `ident.1` and leave
`ident.2 = NULL`.

**NOTE 2**: by default, this function uses a Wilcoxon rank sum test, although other
methods are supported (see `?FindMarkers()` and the Seurat DE vignette referenced
above for help).

## Differential Abundance analysis

Differential Abundance (DA) analysis is designed to detect changes in cell and
cell type composition across conditions in single-cell datasets. To demonstrate
how to perform DA analysis, we will use the `miloR` package. While differential 
abundance (DA) is commonly quantified in discrete cell clusters, Milo uses a 
network of cell-to-cell similarities builg using the k-Nearest Neighbours (KNN)
method, and then analyzes partially overlapping neighbourhoods of cells on the 
KNN graph to test for DA.

For this, we will load a pre-processed version of the Setty et al. dataset in
which single-cell data from **two different individuals**, have been integrated. 
We will therefore test DA between both donors.

**NOTE:** we have added replicate information to the "replicate" column in the
metadata table, generating a replicated multi-condition experiment, which is the
requirement for Milo. Usually, experiments will contain technical replicate or 
batch information associated with sequencing. In this case, however, replicate 
tags were randomly added to cells from each individual to allow for `miloR` to 
be run.

```{r}
setty_combined <- LoadH5Seurat("data_preparation/setty_combined_rep1-rep2.h5seurat",
                               verbose = FALSE)

setty_combined

DimPlot(setty_combined, reduction = "tsne",
        group.by = "predicted.celltype.l2")

DimPlot(setty_combined, reduction = "pca",
        group.by = "individual")
```

Observe how PCA does not separate cells by individual, which is an indicator of
successful integration.

First, we need to transform this to a `SingleCellExperiment` object, which will
then be converted to a `Milo` object, a related class used by the `miloR` package:

```{r}
# build SingleCellExperiment object
combined_sce <- as.SingleCellExperiment(setty_combined, 
                                        assay = "integrated")
combined_sce <- runPCA(combined_sce)

# build Milo object
combined_milo <- Milo(combined_sce)

# remove SingleCellExperiment object to free memory
rm(setty_combined)
```

Next, we need to add the **KNN graph** to the Milo object. This is stored in 
the `graph` slot in the `Milo` object. The miloR package includes functionality 
to build and store the graph from the PCA dimensions stored in the `reducedDim` 
slot.

We will define two parameters:
- `k`: the number of cell nearest-neighbours to use for KNN graph building
- `d`: the number of reduced dimensions to use for KNN graph building

```{r}
combined_milo <- buildGraph(combined_milo, k = 50, d = 50)
```

Now, we will define the neighbourhood of each cell, i.e. the index, which is 
dfined as the group of cells connected by an edge in the KNN graph to the 
index cell. We will set `k` and `d` to the same values as before, and `prop=0.2`,
which corresponds to the proportion of cells to randomly sample to start the
algorithm.

```{r}
combined_milo <- makeNhoods(combined_milo, prop = 0.2,
                            k = 50, d = 50, refined = FALSE)
```

Once we have defined neighbourhoods, it’s good to take a look at how big the 
neighbourhoods are (i.e. how many cells form each neighbourhood). This affects 
the power of DA testing. We can check this out using the plotNhoodSizeHist function. 
Empirically, we found it’s best to have a distribution peaking between 50 and 100. 
Otherwise you might consider rerunning makeNhoods increasing `k` and/or `prop`:

```{r}
plotNhoodSizeHist(combined_milo)
```

**NOTE:** for this tutorial, the `k` and `d` parameters have been adjusted to
generate a distribution that satisifed this indicator. In a real-life scenario,
we would need to adjust the parameters again to generate an adequate size of 
neighbourhoods for our data.

Now we have to count how many cells from each sample are in each neighbourhood. 
We need to use the cell metadata, contained in the `SingleCellExperiment` object,
and specify which column contains the replicate (i.e. sample) information:

```{r}
combined_milo <- countCells(combined_milo,
                            meta.data = data.frame(colData(combined_sce)),
                            samples = "replicate")
```

This adds to the `Milo` object a `n \times m` matrix, where n is the number 
of neighbourhoods and m is the number of experimental samples (in our case, 
the replicates). Values indicate the number of cells from each sample counted 
in a neighbourhood. This count matrix will be used for DA testing.

```{r}
head(nhoodCounts(combined_milo))
```

Now we are all set to test for differential abundance in neighbourhoods. `miloR`
implement this hypothesis testing in a generalized linear model (GLM) framework, 
specifically using the Negative Binomial GLM implementation in `edgeR`.

We first need to think about our experimental design. The design matrix should 
match the replicates to a condition of interest, in this case, the donors defined
in the `individual` column. As a result, we will compare DA across the two donors:

```{r}
design <- data.frame(replicate = colnames(nhoodCounts(combined_milo)),
                     individual = c("Ind1", "Ind1", "Ind2", "Ind2"))
rownames(design) <- design$replicate

design
```

Milo uses a type of False Discovery Rate (FDR) correction that accounts for the
overlap between neighbourhoods, i.e. cells may belong to different neighbourhoods
and therefore be tested multiple times during DA analysis. To use this correction,
we first need to store the distances between nearest neighbors in the `Milo` object
using the `calcNhoodDistance()` function:

```{r, eval = FALSE}
combined_milo <- calcNhoodDistance(combined_milo, d = 50)
```

Since this can be memory intensive, we have stored a versino of the `Milo` object 
with pre-computed distances that you can load:

```{r}
combined_milo <- readRDS("processed_data/combined_milo_NhoodDist.rds")
```


Then, we will do the test using the `testNhoods()` function and our 
experimental design matrix. This calculates a Fold-change and corrected P-value 
for each neighbourhood, which will indicate whether there is significant 
differential abundance between indidivuals.

```{r}
da_results <- testNhoods(combined_milo, 
                         design = ~ individual, 
                         design.df = design)

head(da_results[order(da_results$SpatialFDR, decreasing = TRUE),])
```

Finally, to visualize the DA results, we build an abstracted graph of neighbourhoods
that we can plot together with the single-cell reduced dimension plot:

```{r}
combined_milo <- buildNhoodGraph(combined_milo)

combined_milo <- runUMAP(combined_milo)

# load palettes for visualization
library(paletteer)

plotUMAP(combined_milo, colour_by = "predicted.celltype.l2", point_size = 0.5) + 
  scale_color_manual(values = paletteer_d("ggsci::default_ucscgb"))

plotNhoodGraphDA(combined_milo, da_results, alpha = 0.05)
```

Comparing these plots side by side, we observe that there are particularly 
large abundance changes in the area where pre-B cells accumulate, 
indicating that there are less B cells in individual 2 compared to individual 1.
Similar patterns are observed in yellow-colored cells, which in this case 
correspond to Lympho-Myeloid Pirmed Progenitor (LMPP) cells. All in all, we 
may hypothesize that individual 1 has a larger amount of progenitors that 
differentiate into B/T lymphocytes than individual 2.


